pilot/pkg/proxy/envoy/v2/discovery_test.go

```diff
@@ -295,3 +295,76 @@ func TestDebounce(t *testing.T) {
 		})
 	}
 }
+
+func TestDebounceSyncPush(t *testing.T) {
+	// This test tests the timeout and debouncing of config updates
+	// If it is flaking, DebounceAfter may need to be increased, or the code refactored to mock time.
+	// For now, this seems to work well
+	DebounceAfter = time.Millisecond * 50
+	DebounceMax = DebounceAfter * 2
+
+	tests := []struct {
+		name string
+		test func(updateCh chan *model.PushRequest, expect func(partial, full int32))
+	}{
+		{
+			name: "Should push synchronously after debounce",
+			test: func(updateCh chan *model.PushRequest, expect func(partial, full int32)) {
+				updateCh <- &model.PushRequest{Full: true}
+				time.Sleep(DebounceAfter + 10*time.Millisecond)
+				updateCh <- &model.PushRequest{Full: true}
+				expect(0, 2)
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			stopCh := make(chan struct{})
+			updateCh := make(chan *model.PushRequest, 10)
+			pushingCh := make(chan struct{}, 1)
+
+			var partialPushes int32
+			var fullPushes int32
+
+			fakePush := func(req *model.PushRequest) {
+				select {
+				case pushingCh <- struct{}{}:
+				default:
+					t.Fatalf("multiple pushes happen simutaneously")
```

Comment:
you cannot call t.Fatal inside a goroutine or it will break the tests

Reply:
So have to think of a new error notification

Reply:
To be clear `t.Error` does not work either, see https://golang.org/pkg/testing/#B.FailNow